src/software/util/design_patterns/generic_factory.h

```diff
@@ -0,0 +1,116 @@
+#pragma once
+#include <memory>
+#include <string>
+#include <vector>
+#include <unordered_map>
+#include <functional>
+//#include "software/backend/backend.h" // Should make generic.h
+
+// A quality of life typedef to make things shorter and more readable
+template <class IndexType, class TypeToCreate>
+using GenericRegistry = std::unordered_map<IndexType, std::function<std::unique_ptr<TypeToCreate>()>>;
+/**
+ * The GenericFactory is an Abstract class that provides an interface for Generic type
+ * Factories to follow. This makes it easy to maintain a list of factories and get the
+ * corresponding generic types through the generic interface.
+ */
+template <class IndexType, class TypeToCreate>
+class GenericFactory
+{
+public:
+    /**
+     * Returns a unique pointer to a newly constructed type of the given type/name
+     *
+     * @param generic_name The name of the type to construct. This value must be in the
+     * Generic registry
+     * @throws std::invalid_argument if the given generic_name is not found in the Generic
+     * registry
+     *
+     * @return a unique pointer to a newly constructed type of the given type/name
+     */
+    static std::unique_ptr<TypeToCreate> create(const std::string& generic_name);
+
+    /**
+     * Returns a const reference to the generic type registry. The registry is a map of generic
+     * names to a "create" function that will create and return a unique_ptr to a new
+     * concrete instance of the generic type.
+     *
+     * @return a const reference to the generic registry
+     */
+    static const GenericRegistry<IndexType, TypeToCreate>& getRegistry();
+
+    /**
+     * Returns a list of names of all the existing generic types
+     *
+     * @return a list of names of all the existing generic types
+     */
+    static std::vector<std::string> getRegisteredNames();
+
+    /**
+     * Returns a list of constructor functions for all the existing generic types
```

Comment:
Similar to the above comment, I think this should say something more like "Returns the list of creator functions that are registered in this factory"

Reply:
Done.